### ISSUE TITLE
New distributed cache implementation based on Microsoft SQL Server

### ISSUE THREAD

**Issue by kichalla** [aspnet/Announcements#43]
A new distributed cache implementation based on Microsoft SQL Server was checked in few days back.
 
There are two nuget packages related to this:
- `Microsoft.Framework.Caching.SqlServer` - Has the actual cache implementation
- `Microsoft.Framework.Caching.SqlConfig` - Has the 'global' command to set up the database table and indexes
 
Using this cache requires setting up the database and tables that this implementation understands. Follow the steps below for that:
- Create a database in the SQL server manually.
- Run `dnu commands install Microsoft.Framework.Caching.SqlConfig` to install the command called `sqlservercache`
- Run `sqlservercache create <connectionstring-to-database> <schema-name> <table-name>` to set up the table and indexes.
- Register this cache in DI. Example:
```c#
services.AddSqlServerCache(o =>
{
    o.ConnectionString = "Server=localhost;Database=CacheSampleDb;Trusted_Connection=True;";
    o.SchemaName = "dbo";
    o.TableName = "CacheSample";
});
```
 
And End-to-end example of using this cache is with Session middleware where session data is stored in database making it available across multiple servers behind a load balancer.
```c#
// Registering services
services.AddSqlServerCache(o =>
{
    o.ConnectionString = "Server=localhost;Database=CacheSampleDb;Trusted_Connection=True;";
    o.SchemaName = "dbo";
    o.TableName = "CacheSample";
});
services.AddSession();
 
// Registering session middleware
app.UseSession();
app.UseMvcWithDefaultRoute();
```

Code:
- https://github.com/aspnet/Caching/tree/dev/src/Microsoft.Framework.Caching.SqlServer
- https://github.com/aspnet/Caching/tree/dev/src/Microsoft.Framework.Caching.SqlConfig

Samples:
- https://github.com/aspnet/Caching/tree/dev/samples/SqlServerCacheSample
 
Feedback is welcome

**Comment by tedvanderveen** [comment]
When running an Azure Web app, how is this new distributed cache option positioned against Redis distributed cache in Azure? When choosing between this new cache hosted in SQL Azure or using Azure Redis, what are the main differences?

**Comment by Eilon** [comment]
@tedvanderveen we want to have *both* options! But this SQL Server one is usable easily outside Azure.

**Comment by dauger** [comment]
Has having the option of using Sql Server 2014's in-memory OLTP instead of traditional tables been discussed? It seems like it might be a good fit for a cache, but I'll admit that I have not done a deep dive on in-memory OLTP's capabilities & restrictions yet.

**Comment by Eilon** [comment]
@dauger you can start a discussion in the caching repo. Might be able to gauge interest there.

**Comment by tedvanderveen** [comment]
@dauger that is true if you have an SQL Server instance available (close to your webfarm). Otherwise installation setup of a local Redis instance is easier (and cheaper). Are you planning to support the same rich API as Redis?
Don't get me wrong here, I'm just trying to find out what added value this new distributed cache option brings to the table.

**Comment by RehanSaeed** [comment]
You can use Redis cache like this:

```
services.AddTransient<IDistributedCache, RedisCache>();
```

The code for it is [here](https://github.com/aspnet/Caching/blob/e233e59f9c4bbe7511572264c3c4f092aa0b0a3b/src/Microsoft.Framework.Caching.Redis/RedisCache.cs). Redis is definately the way to go, unless SQL Server 2014 In-Memory caching is possible.

**Comment by kichalla** [comment]
FYI...If you are using RedisCache try to use it as a `Singleton` service as it uses StackExchange's api which recommends sharing the ConnectionMultiplexer 
https://github.com/StackExchange/StackExchange.Redis/blob/master/Docs/Basics.md

I believe the SessionSample that we currently have uses `Transient` which needs to be fixed.

**Comment by PrimeObjects** [comment]
I installed sqlservercache successfully by running dnu commands install Microsoft.Framework.Caching.SqlConfig

By I got error " can't load file or assembly Microsoft.Framework.ApplicationHost or one of dependencies." 
when running  sqlservercache

**Comment by kichalla** [comment]
What is the version of of `dnu` that you are using and the version of `Microsoft.Framework.Caching.SqlConfig` that got installed?

**Comment by PrimeObjects** [comment]
1.0.0-beta7-15019

**Comment by PrimeObjects** [comment]
does that mean I should upgrade all my ASP.NET versions? I'm at beta6

**Comment by kichalla** [comment]
Its ideal to keep working with the version of package and the DNX in sync as breaking changes between different releases might cause issues like you are seeing. So for your scenario make sure you are using Beta6 DNX if you are installing a Beta6 SqlConfig package and try again and let us know.
